feat: remove-x-internal decorator (#443)
Redocly/openapi-cli: docs/resources/built-in-decorators.md

      tagNames:
        pet: ./my-custom-description.md
```

### remove-x-internal

> Using a specific property (`x-internal` by default ) to remove any node.

Reviewer: done

> Remove all nodes that have a specific `internalPropertyName` property (`x-internal` by default ).

Reviewer 2: ```suggestion
Remove all nodes that have a specific `internalPropertyName` property (`x-internal` by default).
```

Reviewer: done

> The `internalPropertyName` option uses to define name of tag.

Reviewer 3: ```suggestion
The `internalPropertyName` option can be used to define the name of the property that is used as flag for removal.
```

```yaml
lint:
  decorators:
    remove-x-internal:
      internalPropertyName: 'removeit'
```

### remove-unused-components

> Remove unused components. We recommend using it with `remove-x-internal` decorator.

Reviewer: done

> Removes unused components.

Reviewer 2: ```suggestion
Removes unused components.
Also can be used as a CLI argument `--remove-unused-components` to the `bundle` command.
```

Reviewer: done

      tagNames:
        pet: ./my-custom-description.md
```

### remove-x-internal

Remove all nodes that have a specific `internalPropertyName` property (`x-internal` by default).

> Remove additional remnants from components by also using the `remove-unused-components` decorator.

Reviewer: done

```yaml
lint:
  decorators:
    remove-x-internal:
      internalPropertyName: 'removeit'
```

### remove-unused-components

Removes unused components.

> Also can be used as a CLI argument `--remove-unused-components` to the `bundle` command.

Reviewer 3: I would not expose it as a decorator as it has to be run last which is a potential place for confusing/unexpected behaviors.

Reviewer: done

      tagNames:
        pet: ./my-custom-description.md
```

> ### hide-internals

Reviewer 2: The opening issue calls for it to be `hide-x-internal` but I wonder if `remove-x-internal` would be more clear.

Reviewer 2: I mention below I see this is more flexible by allowing another token instead of `x-internal` for evaluation of what to remove.

Reviewer: @adamaltman 
`remove-x-internal` sounds good, for now using `hide-x-internal`.
@RomanHotsiy what do you think?

Reviewer 3: `remove-x-internal` sounds better 👍

Reviewer: done

> Using `x-internal` tag by default to hide nodes:

Reviewer 3: ```suggestion
Using a specific property (`x-internal` by default ) to remove nodes:
```

Reviewer: done

```
  PathItem
  Operation
  Schema
  Response
  RequestBody
  Example
  MediaType
  Server
  Callback

>   Parameter

Reviewer 2: We are missing `Tag` too. That should be added. 

We are missing Schema property (this is different from hiding an entire schema -- just hiding a specific property).

Reviewer 3: Removing tags won't remove corresponding operations though (and I'm not sure this would be a good idea).

We should probably add a note to the docs because it may be confusing otherwise.

Reviewer 2: No, we also need to remove tags though. Remember, we do render tags with no operations.

Reviewer 3: What do you mean by "we also need to remove tags though"? Remove removed tag from all the operations that use it? Or remove operations? (I don't think we should remove operations)

```

> The `tagToHide` option uses to define name of tag.

Reviewer 2: I don't understand this sentence or the example.

Reviewer 2: Oh, this is the tag that is evaluated -- default is `x-internal`?

Reviewer 2: I think given this the decorator name `hide-internals` could make sense (or `remove-internals`). However, this particular argument is unclear, because it doesn't just hide that tag, it hides the node/object it is included in such as an operation, property, etc...

Reviewer 2: I think this should not have the word `tag` at all because the word has semantic meaning in OpenAPI.

Reviewer: @adamaltman @RomanHotsiy, guys could you help me to define the name of this option please.

Reviewer 3: I have a few suggestions:

- `propertyName`
- `byPropertyName` (it will have logical flow: remove-internals -> byPropertyName)
- `byProperty` (it will have logical flow: remove-internals -> byProperty)
- `internalPropertyName`

I don't have a strong opinion here. @adamaltman what do you think?

Reviewer 2: I think @andriyl implemented `internalPropertyName` and it seems good.

Reviewer: done

```yaml
lint:
  decorators:
    hide-internals:
      tagToHide: 'hideit'
```

> ### clear-unused-components

Reviewer 2: Is this a decorator or a command line argument or both? Also, I would recommend to rename it.
```suggestion
### remove-unused-components
```

Reviewer: @adamaltman done.

> Remove unused components. A good point to use with `hide-internals` decorator.

Reviewer 2: ```suggestion
Remove unused components. We recommend to use it with `hide-x-internal` decorator.
```

Reviewer: @adamaltman done.
